src/hookintrprtr.h

```diff
@@ -0,0 +1,183 @@
+** reinvoke your hooks!
+*/
+
+struct InterpreterHooks {
+    void (*visitStat)(Stat stat);
+    void (*enterFunction)(Obj func);
+    void (*leaveFunction)(Obj func);
+    void (*registerStat)(Stat stat);
+    const char * hookName;
+};
+
+
+enum { HookCount = 6 };
+
+extern struct InterpreterHooks * activeHooks[HookCount];
+
+Int ActivateHooks(struct InterpreterHooks * hook);
+Int DeactivateHooks(struct InterpreterHooks * hook);
+
+/****************************************************************************
+**
+** A struct to represent the hooks allowed into printing functions
+**
+**
+** This struct represents a list of functions which will be called whenever
+** a statement or expression is printed. They can be used to provide
+** simple customisation of printing. At the moment they are used by
+** profiling.c, to mark statements which have been executed.
+** Look at that code to get an idea how to use them.
+*/
+struct PrintHooks {
+    void (*printStatPassthrough)(Stat stat);
+    void (*printExprPassthrough)(Expr stat);
+    char * hookName;
+};
+
+void ActivatePrintHooks(struct PrintHooks * hook);
+void DeactivatePrintHooks(struct PrintHooks * hook);
+
+/****************************************************************************
+**
+** We can only store a fixed number of files, and files, in statements.
+**
+** Whenever an overflow occurs we do not mark that statement with it's line
+** or file, and call these functions to record the fact information is
+** missing, so we can inform users later.
+*/
+
+void ReportLineNumberOverflowOccured();
+void ReportFileNumberOverflowOccured();
+
+/****************************************************************************
+**
+** We need this to be in the header, so they can be inlined away. The only
+** functionality here which should be publicly used is 'VisitStatIfHooked',
+** 'HookedLineIntoFunction' and 'HookedLineOutFunction'.
+**
+** 'RegisterStatWithHook' is used because some parts of the interpreter
+** skip executing some statements by "cleverness", but we still want them
+** to be visible to code coverage, so they appear 'executed'.
+*/
+
+/* Represents a loop we use frequently. Assumes we have defined
+** an Int i, and InterpreterHooks* hook. We store 'hook' in a local
+** variable to avoid race conditions.
+*/
+
+#define GAP_HOOK_LOOP(member, argument)                                      \
+    for (i = 0; i < HookCount; ++i) {                                        \
+        hook = activeHooks[i];                                               \
+        if (hook && hook->member) {                                          \
+            (hook->member)(argument);                                        \
+        }                                                                    \
+    }
+
+static inline void VisitStatIfHooked(Stat stat)
+{
+    Int                       i;
+    struct InterpreterHooks * hook;
+    GAP_HOOK_LOOP(visitStat, stat);
```

Comment:
Weirdly, I'm sure I tried this, got "not compiled in c99 mode" errors, and took it back out. Now I can't reproduce the c99 complaints I was getting. Will switch to C99 and see if they ever come back.